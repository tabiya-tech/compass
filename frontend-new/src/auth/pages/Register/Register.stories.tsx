import { Meta, StoryObj } from "@storybook/react";
import Register from "./Register";

const meta: Meta<typeof Register> = {
  title: "Auth/Register",
  component: Register,
  tags: ["autodocs"],
};

export default meta;

export const Shown: StoryObj<typeof Register> = {
  args: {},
};
