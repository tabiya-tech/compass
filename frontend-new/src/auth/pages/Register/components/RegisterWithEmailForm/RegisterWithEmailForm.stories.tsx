import { Meta, StoryObj } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import RegisterWithEmailForm from "./RegisterWithEmailForm";

const meta: Meta<typeof RegisterWithEmailForm> = {
  title: "Auth/EmailAuth/RegisterWithEmailForm",
  component: RegisterWithEmailForm,
  tags: ["autodocs"],
};

export default meta;

export const Shown: StoryObj<typeof RegisterWithEmailForm> = {
  args: {
    notifyOnRegister: action("notifyOnRegister"),
    isRegistering: false,
  },
};

export const Registering: StoryObj<typeof RegisterWithEmailForm> = {
  args: {
    notifyOnRegister: action("notifyOnRegister"),
    isRegistering: true,
  },
};
