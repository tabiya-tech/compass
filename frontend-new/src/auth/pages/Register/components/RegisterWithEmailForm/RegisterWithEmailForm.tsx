import { Box, CircularProgress, TextField, useTheme } from "@mui/material";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import React, { useState } from "react";
import { validatePassword } from "src/auth/utils/validatePassword";
import PasswordInput from "src/theme/PasswordInput/PasswordInput";

const uniqueId = "6cf1a0fa-8d75-4342-bf6b-1203d5b114d7";

export const DATA_TEST_ID = {
  FORM: `register-form-${uniqueId}`,
  REGISTRATION_CODE_INPUT: `register-registration-code-input-${uniqueId}`,
  NAME_INPUT: `register-name-input-${uniqueId}`,
  EMAIL_INPUT: `register-email-input-${uniqueId}`,
  PASSWORD_INPUT: `register-password-input-${uniqueId}`,
  REGISTER_BUTTON: `register-button-${uniqueId}`,
  REGISTER_BUTTON_CIRCULAR_PROGRESS: `register-button-circular-progress-${uniqueId}`,
};

export interface RegisterFormProps {
  disabled?: boolean;
  notifyOnRegister: (name: string, email: string, password: string) => void;
  isRegistering: boolean;
}

const RegisterWithEmailForm: React.FC<Readonly<RegisterFormProps>> = ({
  disabled = false,
  notifyOnRegister,
  isRegistering,
}) => {
  const theme = useTheme();
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [passwordError, setPasswordError] = useState<string>("");

  const handleNameChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    setName(event.target.value);
  };
  const handleEmailChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    setEmail(event.target.value);
  };
  const handlePasswordChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    setPassword(event.target.value);
  };

  const handleRegister = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const passwordValidationResult = validatePassword(password);
    setPasswordError(passwordValidationResult);

    if (passwordValidationResult === "") {
      notifyOnRegister(name, email, password);
    }
  };
  return (
    <Box
      display="flex"
      flexDirection="column"
      width={"100%"}
      justifyContent={"space-evenly"}
      gap={theme.fixedSpacing(theme.tabiyaSpacing.md)}
      component="form"
      onSubmit={handleRegister}
      data-testid={DATA_TEST_ID.FORM}
    >
      <TextField
        fullWidth
        label="Name"
        variant="outlined"
        disabled={isRegistering || disabled}
        required
        onChange={(e) => handleNameChange(e)}
        inputProps={{ "data-testid": DATA_TEST_ID.NAME_INPUT }}
      />
      <TextField
        fullWidth
        label="Email"
        type="email"
        disabled={isRegistering || disabled}
        variant="outlined"
        required
        onChange={(e) => handleEmailChange(e)}
        inputProps={{ "data-testid": DATA_TEST_ID.EMAIL_INPUT }}
      />
      <PasswordInput
        fullWidth
        label="Password"
        disabled={isRegistering || disabled}
        variant="outlined"
        required
        onChange={(e) => handlePasswordChange(e)}
        inputProps={{ "data-testid": DATA_TEST_ID.PASSWORD_INPUT }}
        error={!!passwordError}
        helperText={passwordError}
      />
      <PrimaryButton
        fullWidth
        variant="contained"
        color="primary"
        style={{ marginTop: 16 }}
        type="submit"
        disabled={isRegistering || disabled}
        disableWhenOffline={true}
        data-testid={DATA_TEST_ID.REGISTER_BUTTON}
      >
        {isRegistering ? (
          <CircularProgress
            color={"secondary"}
            aria-label={"Registering"}
            data-testid={DATA_TEST_ID.REGISTER_BUTTON_CIRCULAR_PROGRESS}
            size={16}
            sx={{ marginTop: theme.tabiyaSpacing.sm, marginBottom: theme.tabiyaSpacing.sm }}
          />
        ) : (
          "Register"
        )}
      </PrimaryButton>
    </Box>
  );
};

export default RegisterWithEmailForm;
