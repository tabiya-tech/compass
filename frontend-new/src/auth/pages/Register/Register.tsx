import React, { useCallback, useState } from "react";
import { Box, Container, Divider, styled, TextField, Typography, useTheme } from "@mui/material";
import { NavLink as RouterNavLink, useNavigate } from "react-router-dom";
import { routerPaths } from "src/app/routerPaths";
import SocialAuth from "src/auth/components/SocialAuth/SocialAuth";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import RegisterWithEmailForm from "src/auth/pages/Register/components/RegisterWithEmailForm/RegisterWithEmailForm";
import AuthHeader from "src/auth/components/AuthHeader/AuthHeader";
import { FirebaseError, getUserFriendlyFirebaseErrorMessage } from "src/error/FirebaseError/firebaseError";
import { writeFirebaseErrorToLog } from "src/error/FirebaseError/logger";
import { emailAuthService } from "src/auth/services/emailAuth/EmailAuth.service";
import { invitationsService } from "src/invitations/InvitationsService/invitations.service";
import { InvitationStatus, InvitationType } from "src/invitations/InvitationsService/invitations.types";
import { userPreferencesService } from "src/userPreferences/UserPreferencesService/userPreferences.service";
import { Language } from "src/userPreferences/UserPreferencesService/userPreferences.types";
import { getUserFriendlyErrorMessage, ServiceError } from "src/error/ServiceError/ServiceError";
import { writeServiceErrorToLog } from "src/error/ServiceError/logger";
import { logoutService } from "src/auth/services/logout/logout.service";
import { userPreferencesStateService } from "src/userPreferences/UserPreferencesProvider/UserPreferencesStateService";
import { StatusCodes } from "http-status-codes";
import authStateService from "src/auth/AuthStateService";
import { TabiyaUser } from "src/auth/auth.types";
import { Backdrop } from "src/theme/Backdrop/Backdrop";
import FeedbackButton from "src/feedback/FeedbackButton";

const uniqueId = "ab02918f-d559-47ba-9662-ea6b3a3606d0";

const StyledNavLink = styled(RouterNavLink)(({ theme }) => ({
  color: theme.palette.text.textAccent,
  fontStyle: "italic",
  textDecoration: "none",
  "&:hover": {
    textDecoration: "underline",
  },
}));

export const DATA_TEST_ID = {
  REGISTRATION_CODE_INPUT: `register-registration-code-input-${uniqueId}`,
  REGISTER_CONTAINER: `register-container-${uniqueId}`,
  LOGO: `register-logo-${uniqueId}`,
  TITLE: `register-title-${uniqueId}`,
  SUBTITLE: `register-subtitle-${uniqueId}`,
  FORM: `register-form-${uniqueId}`,
  NAME_INPUT: `register-name-input-${uniqueId}`,
  EMAIL_INPUT: `register-email-input-${uniqueId}`,
  PASSWORD_INPUT: `register-password-input-${uniqueId}`,
  REGISTER_BUTTON: `register-button-${uniqueId}`,
  REGISTER_BUTTON_CIRCULAR_PROGRESS: `register-button-circular-progress-${uniqueId}`,
  FORGOT_PASSWORD_LINK: `register-forgot-password-link-${uniqueId}`,
  REGISTER_USING: `register-using-${uniqueId}`,
  FIREBASE_AUTH_CONTAINER: `firebase-auth-container-${uniqueId}`,
  LOGIN_LINK: `register-login-link-${uniqueId}`,
  LANGUAGE_SELECTOR: `register-language-selector-${uniqueId}`,
};

const Register: React.FC = () => {
  const [registrationCode, setRegistrationCode] = useState<string>("");

  const theme = useTheme();

  // a state to determine if the user is currently registering with email
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const { enqueueSnackbar } = useSnackbar();
  const navigate = useNavigate();

  /* -----------
   * callbacks to pass to the child components
   */
  const handleRegistrationCodeChanged = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    setRegistrationCode(event.target.value);
  };

  /**
   * Check if the invitation code is valid
   * @returns {Promise<boolean>} - true if the invitation code is valid, false otherwise
   */
  const isInvitationCodeValid = useCallback(async (): Promise<boolean> => {
    try {
      // Call the service method and use callbacks to determine the result
      const invitation = await invitationsService.checkInvitationCodeStatus(registrationCode);
      if (invitation.status === InvitationStatus.INVALID || invitation.invitation_type !== InvitationType.REGISTER) {
        console.error("Invalid registration code");
        enqueueSnackbar("Invalid registration code", { variant: "error" });
        return false;
      }
      return true;
    } catch (e) {
      let errorMessage;
      if (e instanceof ServiceError) {
        writeServiceErrorToLog(e, console.error);
        errorMessage = getUserFriendlyErrorMessage(e as Error);
      } else {
        console.error(e);
        errorMessage = (e as Error).message;
      }
      enqueueSnackbar(errorMessage, { variant: "error" });
      return false;
    }
  }, [enqueueSnackbar, registrationCode]);

  /**
   * Handles what happens after social registration (same process as login)
   * @param user
   */
  const handlePostLogin = useCallback(
    async (user: TabiyaUser) => {
      try {
        setIsLoading(true);
        const prefs = await userPreferencesService.getUserPreferences(user.id);
        if (prefs === null) {
          throw new Error("User preferences not found");
        }
        userPreferencesStateService.setUserPreferences(prefs);
        if (!prefs?.accepted_tc || isNaN(prefs?.accepted_tc.getTime())) {
          navigate(routerPaths.DPA, { replace: true });
        } else {
          navigate(routerPaths.ROOT, { replace: true });
          enqueueSnackbar("Welcome back!", { variant: "success" });
        }
      } catch (error) {
        let errorMessage;
        if (error instanceof ServiceError) {
          writeServiceErrorToLog(error, console.error);
          errorMessage = getUserFriendlyErrorMessage(error);
        } else {
          errorMessage = (error as Error).message;
          console.error("An error occurred while trying to get your preferences", error);
        }
        enqueueSnackbar(`An error occurred while trying to get your preferences: ${errorMessage}`, {
          variant: "error",
        });
      } finally {
        setIsLoading(false);
      }
    },
    [navigate, enqueueSnackbar]
  );

  /* ------------
   * Actual registration handlers
   */
  /**
   * Handle the register form submission
   * @param name
   * @param email
   * @param password
   */
  const handleRegister = useCallback(
    async (name: string, email: string, password: string) => {
      setIsLoading(true);
      try {
        // check if the invitation code is valid before proceeding with the registration
        const pass = await isInvitationCodeValid();

        if (!pass) {
          setIsLoading(false);
          return;
        }
        const token = await emailAuthService.handleRegisterWithEmail(email, password, name);
        const _user = authStateService.updateUserByToken(token);
        if (_user) {
          // create user preferences for the first time.
          // in order to do this, there needs to be a logged in user in the persistent storage
          const prefs = await userPreferencesService.createUserPreferences({
            user_id: _user.id,
            invitation_code: registrationCode,
            language: Language.en,
          });
          userPreferencesStateService.setUserPreferences(prefs);
          enqueueSnackbar("Verification Email Sent!", { variant: "success" });
          // IMPORTANT NOTE: after the preferences are added, or fail to be added, we should log the user out immediately,
          // since if we don't do that, the user may be able to access the application without verifying their email
          // or accepting the dpa.
          await logoutService.handleLogout();
          userPreferencesStateService.clearUserPreferences();
          await authStateService.clearUser();

          // navigate to the verify email page
          navigate(routerPaths.VERIFY_EMAIL, { replace: true });
        } else {
          // if a user cannot be gotten from the token, we should throw an error
          throw new Error("Something went wrong while registering. Please try again.");
        }
      } catch (e: any) {
        let errorMessage;
        if (e instanceof ServiceError) {
          writeServiceErrorToLog(e, console.error);
          errorMessage = getUserFriendlyErrorMessage(e as Error);
        } else if (e instanceof FirebaseError) {
          writeFirebaseErrorToLog(e, console.error);
          errorMessage = getUserFriendlyFirebaseErrorMessage(e);
        } else {
          console.error(e);
          errorMessage = (e as Error).message;
        }
        enqueueSnackbar(`Registration Failed: ${errorMessage}`, { variant: "error" });
      } finally {
        setIsLoading(false);
      }
    },
    [navigate, enqueueSnackbar, setIsLoading, isInvitationCodeValid, registrationCode]
  );

  /**
   * A callback function for the social auth component to set the loading state
   */
  const notifyOnSocialLoading = useCallback((socialAuthLoading: boolean) => {
    setIsLoading(socialAuthLoading);
  }, []);

  /**
   * A callback to handle what to do after a social registration
   * we pass this to the SocialAuth component which will call it after a successful social registration
   * */
  const successfulSocialRegistrationCallback = async (user: TabiyaUser) => {
    setIsLoading(true);
    try {
      // create user preferences for the first time.
      const prefs = await userPreferencesService.createUserPreferences({
        user_id: user.id,
        invitation_code: registrationCode,
        language: Language.en,
      });
      userPreferencesStateService.setUserPreferences(prefs);
      // We use postLoginHandler because a social registration is parallel to a login
      // and the postLoginHandler is used to handle the post login actions
      await handlePostLogin(user);
    } catch (e: unknown) {
      let errorMessage;
      if (e instanceof ServiceError) {
        // if the user preferences already exist, we should just log the user in
        if (e.statusCode === StatusCodes.CONFLICT) {
          await handlePostLogin(user);
          return;
        }
        errorMessage = getUserFriendlyErrorMessage(e as Error);
        writeServiceErrorToLog(e, console.error);
      } else {
        console.error(e);
        errorMessage = (e as Error).message;
      }
      enqueueSnackbar(`Failed to create preferences: ${errorMessage}`, { variant: "error" });
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <>
      <Container maxWidth="xs" sx={{ height: "100%" }} data-testid={DATA_TEST_ID.REGISTER_CONTAINER}>
        <Box
          display="flex"
          flexDirection="column"
          alignItems="center"
          justifyContent={"space-evenly"}
          gap={theme.fixedSpacing(theme.tabiyaSpacing.sm)}
          height={"80%"}
          width={"100%"}
        >
          <AuthHeader title={"Welcome to Compass!"} subtitle={"We need some information to get started"} />
          <Typography variant="subtitle2">Enter your registration code to sign up</Typography>
          <TextField
            fullWidth
            label="Registration code"
            variant="outlined"
            required
            value={registrationCode}
            onChange={(e) => handleRegistrationCodeChanged(e)}
            inputProps={{ "data-testid": DATA_TEST_ID.REGISTRATION_CODE_INPUT }}
          />
          <Divider textAlign="center" style={{ width: "100%" }}>
            <Typography variant="subtitle2" padding={theme.fixedSpacing(theme.tabiyaSpacing.sm)}>
              and then continue with
            </Typography>
          </Divider>
          <RegisterWithEmailForm
            disabled={!registrationCode}
            notifyOnRegister={handleRegister}
            isRegistering={isLoading}
          />
          <SocialAuth
            preLoginCheck={isInvitationCodeValid}
            postLoginHandler={successfulSocialRegistrationCallback}
            isLoading={isLoading}
            disabled={!registrationCode}
            label={"Sign up with Google"}
            notifyOnLoading={notifyOnSocialLoading}
          />
          <Typography variant="body2" data-testid={DATA_TEST_ID.LOGIN_LINK}>
            Already have an account?{" "}
            <StyledNavLink
              to={routerPaths.LOGIN}
              style={{
                color: theme.palette.text.textAccent,
                fontStyle: "italic",
              }}
            >
              Login
            </StyledNavLink>
          </Typography>
        </Box>
      </Container>
      <FeedbackButton bottomAlign={true} />
      <Backdrop isShown={isLoading} message="Registering you..." />
    </>
  );
};

export default Register;
