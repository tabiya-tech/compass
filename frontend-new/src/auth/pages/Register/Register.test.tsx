import "src/_test_utilities/consoleMock";
import React from "react";
import { render, screen, waitFor, act, fireEvent } from "src/_test_utilities/test-utils";
import { HashRouter } from "react-router-dom";
import Register, { DATA_TEST_ID } from "./Register";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { mockUseTokens } from "src/_test_utilities/mockUseTokens";
import RegisterWithEmailForm from "src/auth/pages/Register/components/RegisterWithEmailForm/RegisterWithEmailForm";
import { DATA_TEST_ID as AUTH_HEADER_DATA_TEST_ID } from "src/auth/components/AuthHeader/AuthHeader";
import { AuthContext, AuthContextValue, TabiyaUser } from "src/auth/AuthProvider";
import { emailAuthService } from "src/auth/services/emailAuth/EmailAuth.service";
import { invitationsService } from "src/invitations/InvitationsService/invitations.service";
import { InvitationStatus, InvitationType } from "src/invitations/InvitationsService/invitations.types";
import { ServiceError } from "src/error/ServiceError/ServiceError";

//mock the SocialAuth component
jest.mock("src/auth/components/SocialAuth/SocialAuth", () => {
  const actual = jest.requireActual("src/auth/components/SocialAuth/SocialAuth");
  return {
    ...actual,
    __esModule: true,
    default: jest.fn().mockImplementation(() => {
      return <span data-testid={actual.DATA_TEST_ID.FIREBASE_AUTH_CONTAINER}></span>;
    }),
  };
});

// mock the emailAuthService
jest.mock("src/auth/services/emailAuth/EmailAuth.service", () => {
  return {
    emailAuthService: {
      handleRegisterWithEmail: jest.fn(),
      handleLogout: jest.fn(),
    },
  };
});
// mock the snack bar provider
jest.mock("src/theme/SnackbarProvider/SnackbarProvider", () => {
  const actual = jest.requireActual("src/theme/SnackbarProvider/SnackbarProvider");
  return {
    ...actual,
    __esModule: true,
    useSnackbar: jest.fn().mockReturnValue({
      enqueueSnackbar: jest.fn(),
      closeSnackbar: jest.fn(),
    }),
  };
});

// mock the router
jest.mock("react-router-dom", () => {
  const actual = jest.requireActual("react-router-dom");
  return {
    ...actual,
    __esModule: true,
    useNavigate: jest.fn().mockReturnValue(jest.fn()),
    NavLink: jest.fn().mockImplementation(() => {
      return <></>;
    }),
  };
});

// mock the RegisterWithEmailForm component
jest.mock("src/auth/pages/Register/components/RegisterWithEmailForm/RegisterWithEmailForm", () => {
  const actual = jest.requireActual("src/auth/pages/Register/components/RegisterWithEmailForm/RegisterWithEmailForm");
  return {
    ...actual,
    __esModule: true,
    default: jest.fn().mockImplementation(() => {
      return <span data-testid={actual.DATA_TEST_ID.FORM}></span>;
    }),
  };
});

// mock the AuthHeader component
jest.mock("src/auth/components/AuthHeader/AuthHeader", () => {
  const actual = jest.requireActual("src/auth/components/AuthHeader/AuthHeader");
  return {
    ...actual,
    __esModule: true,
    default: jest.fn().mockImplementation(() => {
      return <span data-testid={actual.DATA_TEST_ID.AUTH_HEADER_CONTAINER}></span>;
    }),
  };
});

describe("Testing Register component", () => {
  const authContextValue: AuthContextValue = {
    user: null,
    updateUserByToken: jest.fn(),
    clearUser: jest.fn(),
    isAuthenticationInProgress: false,
    isAuthenticated: false,
  };

  beforeEach(() => {
    // Clear console mocks and mock functions
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
    jest.clearAllMocks();
  });

  beforeAll(() => mockUseTokens());

  test("it should show register form successfully", async () => {
    // GIVEN a user to register
    const givenInvitationCode = "foo-bar";
    const givenName = "Foo Bar";
    const givenEmail = "foo@bar.baz";
    const givenPassword = "password";
    const givenNotifyOnRegister = jest.fn();
    const givenNotifyOnLogin = jest.fn();
    const givenIsLoading = false;

    // AND check invitation code status returns a valid code
    const checkInvitationCodeStatusMock = jest
      .spyOn(invitationsService, "checkInvitationCodeStatus")
      .mockImplementation(
        //@ts-ignore
        (initation_code, successCallback, failureCallback) => {
          successCallback({
            invitation_type: InvitationType.REGISTER,
            status: InvitationStatus.VALID,
            invitation_code: initation_code,
          });
        }
      );
    // WHEN the component is rendered within the AuthContext and Router
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <Register
            postRegisterHandler={givenNotifyOnRegister}
            postLoginHandler={givenNotifyOnLogin}
            isPostLoginLoading={givenIsLoading}
          />
        </AuthContext.Provider>
      </HashRouter>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND the component should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.REGISTER_CONTAINER)).toBeInTheDocument();

    // AND the header component should be rendered
    expect(screen.getByTestId(AUTH_HEADER_DATA_TEST_ID.AUTH_HEADER_CONTAINER)).toBeInTheDocument();

    // AND the form inputs and button should be displayed
    expect(RegisterWithEmailForm).toHaveBeenCalled();

    fireEvent.change(screen.getByTestId(DATA_TEST_ID.REGISTRATION_CODE_INPUT), {
      target: { value: givenInvitationCode },
    });

    await act(async () => {
      // Simulate form submission
      const calls = (RegisterWithEmailForm as jest.Mock).mock.calls;

      await calls[calls.length - 1][0].notifyOnRegister(givenName, givenEmail, givenPassword);
    });

    await waitFor(() => {
      expect(screen.getByTestId(DATA_TEST_ID.REGISTRATION_CODE_INPUT)).toHaveValue(givenInvitationCode);
    });

    // Expect the register function to have been called
    await waitFor(() => {
      expect(emailAuthService.handleRegisterWithEmail).toHaveBeenCalledWith(
        givenEmail,
        givenPassword,
        givenName,
        expect.any(Function),
        expect.any(Function)
      );
    });

    // AND check that the invitation code status was checked
    expect(checkInvitationCodeStatusMock).toHaveBeenCalledWith(
      givenInvitationCode,
      expect.any(Function),
      expect.any(Function)
    );

    // AND the component should match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.REGISTER_CONTAINER)).toMatchSnapshot();
  });

  test("it should show success message on successful registration", async () => {
    // GIVEN a successful registration
    const givenName = "Foo Bar";
    const givenEmail = "foo@bar.baz";
    const givenPassword = "password";
    const givenNotifyOnRegister = jest.fn();
    const givenNotifyOnLogin = jest.fn();
    const givenIsLoading = false;

    (emailAuthService.handleRegisterWithEmail as jest.Mock).mockImplementation(
      (
        email: string,
        password: string,
        name: string,
        onSuccess: (user: TabiyaUser) => void,
        onError: (error: Error) => void
      ) => {
        onSuccess({ id: "mock-id", email: givenEmail, name: givenName } as TabiyaUser);
      }
    );

    // WHEN the component is rendered
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <Register
            postRegisterHandler={givenNotifyOnRegister}
            postLoginHandler={givenNotifyOnLogin}
            isPostLoginLoading={givenIsLoading}
          />
        </AuthContext.Provider>
      </HashRouter>
    );

    // AND the register form is submitted
    await act(() => {
      (RegisterWithEmailForm as jest.Mock).mock.calls[0][0].notifyOnRegister(givenName, givenEmail, givenPassword);
    });

    await act(async () => {
      // Simulate form submission
      const calls = (RegisterWithEmailForm as jest.Mock).mock.calls;

      await calls[calls.length - 1][0].notifyOnRegister(givenName, givenEmail, givenPassword);
    });

    // THEN expect the register function to have been called
    await waitFor(() => {
      expect(emailAuthService.handleRegisterWithEmail).toHaveBeenCalledWith(
        givenEmail,
        givenPassword,
        givenName,
        expect.any(Function),
        expect.any(Function)
      );
    });

    // AND no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
  });

  test("it should show error message on failed registration", async () => {
    // GIVEN a failed registration
    (emailAuthService.handleRegisterWithEmail as jest.Mock).mockImplementation(
      (
        email: string,
        password: string,
        name: string,
        onSuccess: (user: TabiyaUser) => void,
        onError: (error: Error) => void
      ) => {
        onError(new Error("An unexpected error occurred. Please try again later."));
      }
    );

    const givenName = "Foo Bar";
    const givenEmail = "foo@bar.baz";
    const givenPassword = "password";
    const givenNotifyOnRegister = jest.fn();
    const givenNotifyOnLogin = jest.fn();
    const givenIsLoading = false;

    // AND check invitation code status returns a valid code
    jest.spyOn(invitationsService, "checkInvitationCodeStatus").mockImplementation(
      //@ts-ignore
      (initation_code, successCallback, failureCallback) => {
        successCallback({
          invitation_type: InvitationType.REGISTER,
          status: InvitationStatus.VALID,
          invitation_code: initation_code,
        });
      }
    );

    // WHEN the register form is submitted
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <Register
            postRegisterHandler={givenNotifyOnRegister}
            postLoginHandler={givenNotifyOnLogin}
            isPostLoginLoading={givenIsLoading}
          />
        </AuthContext.Provider>
      </HashRouter>
    );

    // AND the register form is submitted
    await act(() => {
      (RegisterWithEmailForm as jest.Mock).mock.calls[0][0].notifyOnRegister(givenName, givenEmail, givenPassword);
    });
    // THEN expect the register function to have been called
    await waitFor(() => {
      expect(emailAuthService.handleRegisterWithEmail).toHaveBeenCalledWith(
        givenEmail,
        givenPassword,
        givenName,
        expect.any(Function),
        expect.any(Function)
      );
    });

    // AND the error message should be displayed
    expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith("An internal error has occurred.", { variant: "error" });
  });

  test("it should not call register when invitation code is invalid", async () => {
    // GIVEN an invalid invitation code
    const givenInvitationCode = "foo-bar";
    const givenName = "Foo Bar";
    const givenEmail = "foo-bar@foo.bar";

    // AND check invitation code check returns an invalid code
    jest.spyOn(invitationsService, "checkInvitationCodeStatus").mockImplementation(
      //@ts-ignore
      (initation_code, successCallback, failureCallback) => {
        failureCallback({
          message: "Invalid invitation code",
        } as ServiceError);
      }
    );

    // WHEN the component is rendered
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <Register postRegisterHandler={jest.fn()} postLoginHandler={jest.fn()} isPostLoginLoading={false} />
        </AuthContext.Provider>
      </HashRouter>
    );

    fireEvent.change(screen.getByTestId(DATA_TEST_ID.REGISTRATION_CODE_INPUT), {
      target: { value: givenInvitationCode },
    });

    // AND the register form is submitted
    (RegisterWithEmailForm as jest.Mock).mock.calls[0][0].notifyOnRegister(givenName, givenEmail, "password");

    // THEN expect the register function to not have been called
    await waitFor(() => {
      expect(emailAuthService.handleRegisterWithEmail).not.toHaveBeenCalled();
    });
  });
});
