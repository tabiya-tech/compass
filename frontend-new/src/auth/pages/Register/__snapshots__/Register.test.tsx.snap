// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Register component it should show register form successfully 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthXs css-xravbe-MuiContainer-root"
  data-testid="register-container-ab02918f-d559-47ba-9662-ea6b3a3606d0"
>
  <div
    class="MuiBox-root css-i9eu8x"
  >
    <span
      data-testid="auth-header-container-40374529-6e2c-49d8-81d7-93f01603a648"
    />
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-ohbpz3-MuiTypography-root"
    >
      Enter your registration code to sign up
    </h6>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1pcwqdd-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1d7vtq4-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Registration code
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1th3aq5-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1vdto9g-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-kqqmrr-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="register-registration-code-input-ab02918f-d559-47ba-9662-ea6b3a3606d0"
          id=":r0:"
          required=""
          type="text"
          value="foo-bar"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Registration code
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1puturl-MuiDivider-root"
      role="separator"
      style="width: 100%;"
    >
      <span
        class="MuiDivider-wrapper css-8j2o7p-MuiDivider-wrapper"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle2 css-zq12aa-MuiTypography-root"
        >
          and then continue with
        </h6>
      </span>
    </div>
    <span
      data-testid="register-form-6cf1a0fa-8d75-4342-bf6b-1203d5b114d7"
    />
    <span
      data-testid="firebase-auth-container-f0324e97-83fd-49e6-95c3-1043751fa1db"
    />
    <p
      class="MuiTypography-root MuiTypography-body2 css-bbtak5-MuiTypography-root"
      data-testid="register-login-link-ab02918f-d559-47ba-9662-ea6b3a3606d0"
    >
      Already have an account?
       
    </p>
  </div>
</div>
`;
